### ISSUE TITLE
feat(tools): Allow multiple arb files to organize l10n / intl localizations

### ISSUE THREAD

**Issue by golane-august** [flutter/flutter#107157]
## Use case

In the current app development, you cannot split your localizations into multiple files to better categorize them. You also don't have the ability to add comments to JSON files in order to better separate the categories.
In Flutter itself you also have [multiple files](https://github.com/flutter/flutter/tree/master/packages/flutter_localizations/lib/src/l10n) for the different packages.

See also docs for [localization](https://docs.flutter.dev/development/accessibility-and-localization/internationalization).

Question is also [asked on Stackoverflow](https://stackoverflow.com/questions/65440222/organise-arb-files-in-flutter-using-flutter-intl-extension) with no good answer yet.

Dart package is claimed to support multiple arb files: https://pub.dev/packages/intl_translation

## Proposal

- Allow multiple `.arb` files with the same suffix
- 
The relevant [changes could be made here](https://github.com/flutter/flutter/blob/f3717d70d7d2660ed830835c8ee33a01af71ce08/packages/flutter_tools/lib/src/localizations/gen_l10n_types.dart#L532).

**Comment by maheshmnj** [comment]
Hi @golane-august, Thanks for filing the issue. That looks like a valid feature request.

**Comment by sbatezat** [comment]
It's also a must have from my point of view.
Having all localized strings into one file make things very complicated.

However, I think things could be even more improved than @golane-august suggestion.
Having all files into a unique folder is not suitable from my point of view.
Basically, we should have the ability to store arb files near to each linked widget.

Example of app structure I have in mind:

```
lib/features
   home/
       homeWidget.dart
       home_en.arb
       home_fr.arb
       ....
   awesomeFeature
       awesomeFeatureWidget.dart
       awesomeFeature_en.arb
       awesomeFeature_fr.arb
```

If needed, we may use special attributes of arb files, **@@locale** and **@@context** described at https://github.com/google/app-resource-bundle/wiki/ApplicationResourceBundleSpecification

**Comment by golane-august** [comment]
@sbatezat I don't really agree on that. Although it would be nice to have more flexibility, it may is a bad idea everyone cooks their own soup. I more like the idea of having reusable strings (e.g. such as for buttons like `OK` or `Cancel`), which doesn't belong to any special widget. Also in `src` only should be source files and no translations. You also usually don't mix up resources or assets into source folder. If it should be really separated it's worth considering writing an extra package for this "widget" or functionality.

**Comment by ziqq** [comment]
> It's also a must have from my point of view. Having all localized strings into one file make things very complicated.
> 
> However, I think things could be even more improved than @golane-august suggestion. Having all files into a unique folder is not suitable from my point of view. Basically, we should have the ability to store arb files near to each linked widget.
> 
> Example of app structure I have in mind:
> 
> ```
> lib/features
>    home/
>        homeWidget.dart
>        home_en.arb
>        home_fr.arb
>        ....
>    awesomeFeature
>        awesomeFeatureWidget.dart
>        awesomeFeature_en.arb
>        awesomeFeature_fr.arb
> ```
> 
> If needed, we may use special attributes of arb files, **@@ locale** and **@@ context** described at https://github.com/google/app-resource-bundle/wiki/ApplicationResourceBundleSpecification
> 
> See also this issue/comment, showing we are not the only ones interested by cleaning our arb files : [#71729 (comment)](https://github.com/flutter/flutter/issues/71729#issuecomment-852059076)

I want to make this.

**Comment by joeldegerman99** [comment]
Agree, would be much easier to maintain if we could split it up into feature specific arb files

**Comment by ricardodalarme** [comment]
up

**Comment by vscmaster** [comment]
On more thumb-up to get localisation gens separated. 
We are facing an issue with long file at the moment, that is difficult to read more than 1k lines per file.

Could you manage to generate Localisation per feature(arb file).

**Comment by KPatel91** [comment]
We would also like something like this, akin to .NET resource files. We have quite a large app with multiple flavours, for different customers (of which some want multi-lingual, others don't) - and would like to be able to have feature-specific resources, flavour-specific resources, and just a bit of organisation. 

Otherwise we end up with thousands of fields in a single file and have to manage namespacing the keys ourselves.

Would be quite nice even just to be able to simply have multiple .arb files and a parent file that lists which ones to include in the release.

**Comment by edokichuks** [comment]
> 

This really needs to be done as it will be difficult to maintain the translation without it

**Comment by ricardodalarme** [comment]
Hey everyone, I've started working on this and already created a draft. You can check it out there. The part that generates the localizations with a namespace is already working. I plan to finish it by the end of next week.
- https://github.com/flutter/flutter/pull/137307

**Comment by geminiyellow** [comment]
good, hope we can use it in early 2024.

**Comment by ndrslmpk** [comment]
Any updates on this?

**Comment by cedvdb** [comment]
> Any updates on this?

The PR is stale. Someone can take it over https://github.com/flutter/flutter/pull/137307

**Comment by cedvdb** [comment]
While waiting I've been using https://pub.dev/packages/merge_localizations

**Comment by klamberdiereITLink** [comment]
One more. 
This kind of feature would greatly help for having I18N by flavors.

**Comment by ricardodalarme** [comment]
I just opened a new PR for this feature, now with backward compatibility:

- https://github.com/flutter/flutter/pull/157696

I'd love it if you could give a review and add suggestions :)

**Comment by hpelitebook745G2** [comment]
If anyone has the same use case as me, I found this [article](https://medium.com/@alexey.yu.popkov/flutter-tip-organize-your-localization-with-multiple-arb-files-per-language-52606e5dbf25) and I hope this helps.

tl;dr I wanted to have multiple `.arb` file sources:
1. Place your `<name>.arb` files in separate folders:

```
- l10n
  - app
    - app_en.arb
  - errors
    - errors_en.arb
 ```
2. Create a shell script, `generate-l10n.sh`:
```
#!/bin/bash

flutter gen-l10n \
  --arb-dir ./lib/src/localization/app \
  --template-arb-file app_en.arb \
  --output-localization-file app_localizations.dart \
  --output-class AppLocalizations \
  || { echo "Error generating app l10n"; exit 1; }

flutter gen-l10n \
  --arb-dir ./lib/src/localization/errors \
  --template-arb-file errors_en.arb \
  --output-localization-file errors_localizations.dart \
  --output-class ErrorsLocalizations \
  || { echo "Error generating error messages l10n"; exit 1; }
```
Run the script: `sh ./generate-l10n.sh`
3. If you have existing `l10n.yaml`, delete it as the script will locate your `.arb` files
4. In your `MaterialApp`, add the the delegates:
```
MaterialApp.router(
  builder: (context, widget) {
    return ResponsiveWrapper(
      child: widget!,
    );
  },
  restorationScopeId: 'app',
  localizationsDelegates: const [
    AppLocalizations.delegate, // Add this line
    ErrorsLocalizations.delegate, // Add this line
    GlobalMaterialLocalizations.delegate,
    GlobalWidgetsLocalizations.delegate,
    GlobalCupertinoLocalizations.delegate,
  ],
  supportedLocales: const [
    Locale('en', ''),
  ],
  onGenerateTitle: (BuildContext context) =>
      AppLocalizations.of(context)!.appTitle,
  theme: AffinidiThemeData().light,
  darkTheme: AffinidiThemeData().dark,
  themeMode: ThemeMode.light,
  routerConfig: AppRouter.router,
);
```

**Comment by Piinks** [comment]
There was a PR out to resolve this, but unfortunately was not finished. If anyone would like to tackle this, https://github.com/flutter/flutter/pull/157696 would probably give a good head start. :)

**Comment by JHubi1** [comment]
Maybe it would be better if we made the local identifier the root folder and then in that folder allowed multiple ARB files.

Something like this:
```
l10n/
├── en/
│   ├── home.arb
│   ├── login.arb
│   └── app.arb
└── de/
    ├── home.arb
    ├── login.arb
    └── app.arb
```

This would create a separation between the different languages and might make the experience better for translators. You'd only have to send or upload the source language folder, and the translator or translation tool would only create the localized folder.

Now arguably, without using a tool to manage localization, this approach would probably make it harder to keep track of each file.

In my opinion, this would make a better folder structure. Thoughts?

**Comment by sbatezat** [comment]
> Maybe it would be better if we made the local identifier the root folder and then in that folder allowed multiple ARB files.
> 
> Something like this:
> 
> ```
> l10n/
> ├── en/
> │   ├── home.arb
> │   ├── login.arb
> │   └── app.arb
> └── de/
>     ├── home.arb
>     ├── login.arb
>     └── app.arb
> ```
> 
> This would create a separation between the different languages and might make the experience better for translators. You'd only have to send or upload the source language folder, and the translator or translation tool would only create the localized folder.
> 
> Now arguably, without using a tool to manage localization, this approach would probably make it harder to keep track of each file.
> 
> In my opinion, this would make a better folder structure. Thoughts?

It's the way some framework are working, like Android.

Pros is to group all l10n strings into one dedicated folder (to share to your translation team - for example)
Cons is you don't have "removeable feature". When you delete a feature, you need to think about its dependencies outside the feature folder.

It's all about: do you prefer technical folder structure or functional folder structure?

**Comment by cedvdb** [comment]
@sbatezat You can also have the .arb files in feature folders, which combine in a single file that you can share.
